File: src/solution/s0004_median_of_two_sorted_arrays.rb
# @param {Integer[]} nums1
# @param {Integer[]} nums2
# @return {Float}
def find_median_sorted_arrays(nums1, nums2)
  size = nums1.size + nums2.size
  guard = size / 2
  prev = nil
  current = nil
  i = 0
  j = 0
  while i + j <= guard do
    # puts "prev:#{prev}, current:#{current}, pos:#{i + j}, guard:#{guard}"
    n1 = nums1[i]
    n2 = nums2[j]
    prev = current || 0
    unless n1
      j += 1
      current = n2
      next
    end
    unless n2
      i += 1
      current = n1
      next
    end
    if n1 <= n2
      i += 1
      current = n1
    else
      j += 1
      current = n2
    end
  end
  # puts "  prev:#{prev}, current:#{current}"
  size.even? ? (prev + current) / 2.0 : current.to_f
end

require "rspec/expectations"
include RSpec::Matchers

expect(find_median_sorted_arrays([1, 3], [2])).to eq 2.0
expect(find_median_sorted_arrays([1, 2], [3, 4])).to eq 2.5
expect(find_median_sorted_arrays([0, 0], [0, 0])).to eq 0.0
expect(find_median_sorted_arrays([], [1])).to eq 1.0
expect(find_median_sorted_arrays([2], [])).to eq 2.0

File: src/solution/s0004_median_of_two_sorted_arrays.rs
/**
 * [4] Median of Two Sorted Arrays
 *
 * Given two sorted arrays nums1 and nums2 of size m and n respectively, return the median of the two sorted arrays.
 * Follow up: The overall run time complexity should be O(log (m+n)).
 *
 * Example 1:
 *
 * Input: nums1 = [1,3], nums2 = [2]
 * Output: 2.00000
 * Explanation: merged array = [1,2,3] and median is 2.
 *
 * Example 2:
 *
 * Input: nums1 = [1,2], nums2 = [3,4]
 * Output: 2.50000
 * Explanation: merged array = [1,2,3,4] and median is (2 + 3) / 2 = 2.5.
 *
 * Example 3:
 *
 * Input: nums1 = [0,0], nums2 = [0,0]
 * Output: 0.00000
 *
 * Example 4:
 *
 * Input: nums1 = [], nums2 = [1]
 * Output: 1.00000
 *
 * Example 5:
 *
 * Input: nums1 = [2], nums2 = []
 * Output: 2.00000
 *
 *
 * Constraints:
 *
 * 	nums1.length == m
 * 	nums2.length == n
 * 	0 <= m <= 1000
 * 	0 <= n <= 1000
 * 	1 <= m + n <= 2000
 * 	-10^6 <= nums1[i], nums2[i] <= 10^6
 *
 */
pub struct Solution {}

// problem: https://leetcode.com/problems/median-of-two-sorted-arrays/
// discuss: https://leetcode.com/problems/median-of-two-sorted-arrays/discuss/?currentPage=1&orderBy=most_votes&query=

// submission codes start here

impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let size: usize = nums1.len() + nums2.len();
        let guard: usize = size / 2;
        let mut prev: i32 = 0;
        let mut current: i32 = 0;
        let mut i: usize = 0;
        let mut j: usize = 0;
        while i + j <= guard {
            let n1 = nums1.get(i);
            let n2 = nums2.get(j);
            prev = current;
            if n1 == None {
                j += 1;
                current = *n2.unwrap();
                continue;
            }
            if n2 == None {
                i += 1;
                current = *n1.unwrap();
                continue;
            }
            if *n1.unwrap() <= *n2.unwrap() {
                i += 1;
                current = *n1.unwrap();
            } else {
                j += 1;
                current = *n2.unwrap();
            }
        }
        if size % 2 == 0 {
            (prev + current) as f64 / 2.0
        } else {
            current as f64
        }
    }
}

// submission codes end

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_4() {
        assert_eq!(
            Solution::find_median_sorted_arrays(vec![1, 3], vec![2]),
            2.0
        );
        assert_eq!(
            Solution::find_median_sorted_arrays(vec![1, 2], vec![3, 4]),
            2.5
        );
        assert_eq!(
            Solution::find_median_sorted_arrays(vec![0, 0], vec![0, 0]),
            0.0
        );
        assert_eq!(Solution::find_median_sorted_arrays(vec![], vec![1]), 1.0);
        assert_eq!(Solution::find_median_sorted_arrays(vec![2], vec![]), 2.0);
    }
}
